# Training a model on Number recognition with MNIST dataset

### This training includes different cases which includes 


### Ranging number of neurons in hidden layer

*italicized text*### Different epochs

### Different activationfunctions such ReLU, softmax, sigmoid, tanh

### Different optimizers such as Adam, SGD, Adamax

In [ ]:
import tensorflow as tf
from tensorflow import keras

(x_train, y_train),(x_test, y_test) = keras.datasets.mnist.load_data()

11493376/11490434 [==============================] - 0s 0us/step


In [ ]:
model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(128, activation=tf.nn.relu),
    keras.layers.Dense(10, activation=tf.nn.softmax)
    ])

Instructions for updating:
If using Keras pass *_constraint arguments to layers.


In [ ]:
model.compile(optimizer=tf.train.AdamOptimizer(),
             loss='sparse_categorical_crossentropy',
             metrics=['accuracy']
             )

In [ ]:
model.fit(x_train, y_train, epochs=7)

Train on 60000 samples
Epoch 1/7
60000/60000 [==============================] - 4s 69us/sample - loss: 2.7375 - acc: 0.8632
Epoch 2/7
60000/60000 [==============================] - 4s 62us/sample - loss: 0.3666 - acc: 0.9129
Epoch 3/7
60000/60000 [==============================] - 4s 62us/sample - loss: 0.2837 - acc: 0.9279
Epoch 4/7
60000/60000 [==============================] - 4s 63us/sample - loss: 0.2476 - acc: 0.9378
Epoch 5/7
60000/60000 [==============================] - 4s 64us/sample - loss: 0.2283 - acc: 0.9426
Epoch 6/7
60000/60000 [==============================] - 4s 62us/sample - loss: 0.2141 - acc: 0.9461
Epoch 7/7
60000/60000 [==============================] - 4s 62us/sample - loss: 0.2149 - acc: 0.9474


In [ ]:
test_loss, test_acc = model.evaluate(x_test, y_test)

10000/10000 [==============================] - 0s 41us/sample - loss: 0.3188 - acc: 0.9446


In [ ]:
print("test accuracy : ", test_acc)

test accuracy :  0.9446


In [ ]:
predictions = model.predict(x_test)

## In the first test with 128 nueron in hiddenlayer and with 5 epochs we got 94.00 accuracy

### 2nd test

### testing the model with the same paramets and epochs = 8

In [ ]:
model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(128, activation=tf.nn.relu),
    keras.layers.Dense(10, activation=tf.nn.softmax)
    ])

model.compile(optimizer=tf.train.AdamOptimizer(),
             loss='sparse_categorical_crossentropy',
             metrics=['accuracy']
             )

model.fit(x_train, y_train, epochs=6)
test_loss, test_acc = model.evaluate(x_test, y_test)

Train on 60000 samples
Epoch 1/6
60000/60000 [==============================] - 4s 74us/sample - loss: 2.1735 - acc: 0.8558
Epoch 2/6
60000/60000 [==============================] - 4s 73us/sample - loss: 0.3649 - acc: 0.9110
Epoch 3/6
60000/60000 [==============================] - 5s 76us/sample - loss: 0.2726 - acc: 0.9284
Epoch 4/6
60000/60000 [==============================] - 5s 77us/sample - loss: 0.2531 - acc: 0.9348
Epoch 5/6
60000/60000 [==============================] - 5s 77us/sample - loss: 0.2251 - acc: 0.9424
Epoch 6/6
10000/10000 [==============================] - 0s 38us/sample - loss: 0.2395 - acc: 0.9411


In [ ]:
print("test accuracy : ", test_acc)
predictions = model.predict(x_test)

test accuracy :  0.9487


### with increase in epochs there is a slight increase in the accuracy

### 3rd Test -> testing with half of the nuerons i.e 64

In [ ]:
model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(64, activation=tf.nn.relu),
    keras.layers.Dense(10, activation=tf.nn.softmax)
    ])

model.compile(optimizer=tf.train.AdamOptimizer(),
             loss='sparse_categorical_crossentropy',
             metrics=['accuracy']
             )

model.fit(x_train, y_train, epochs=7)
test_loss, test_acc = model.evaluate(x_test, y_test)

Train on 60000 samples
Epoch 1/7
60000/60000 [==============================] - 4s 69us/sample - loss: 1.8168 - acc: 0.7765
Epoch 2/7
60000/60000 [==============================] - 4s 68us/sample - loss: 0.4313 - acc: 0.8859
Epoch 3/7
60000/60000 [==============================] - 4s 67us/sample - loss: 0.3018 - acc: 0.9208
Epoch 4/7
60000/60000 [==============================] - 4s 68us/sample - loss: 0.2700 - acc: 0.9294
Epoch 5/7
60000/60000 [==============================] - 4s 66us/sample - loss: 0.2512 - acc: 0.9334
Epoch 6/7
60000/60000 [==============================] - 4s 66us/sample - loss: 0.2401 - acc: 0.9370
Epoch 7/7
10000/10000 [==============================] - 0s 36us/sample - loss: 0.2769 - acc: 0.9353


In [ ]:
predictions = model.predict(x_test)
print("test accuracy : ", test_acc)

test accuracy :  0.931


### With decrease in the neurons there is a decrease in the accuracy from 0.9487 to 0.931

### 4th test -> testing with the double the neurons in the hidden layer i.e 180

In [ ]:
model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(64, activation=tf.nn.relu),
    keras.layers.Dense(10, activation=tf.nn.softmax)
    ])

model.compile(optimizer=tf.train.AdamOptimizer(),
             loss='sparse_categorical_crossentropy',
             metrics=['accuracy']
             )

model.fit(x_train, y_train, epochs=5)
test_loss, test_acc = model.evaluate(x_test, y_test)


Train on 60000 samples
Epoch 1/5
60000/60000 [==============================] - 4s 69us/sample - loss: 2.2684 - acc: 0.7820
Epoch 2/5
60000/60000 [==============================] - 4s 71us/sample - loss: 0.4911 - acc: 0.8788
Epoch 3/5
60000/60000 [==============================] - 4s 67us/sample - loss: 0.3520 - acc: 0.9099
Epoch 4/5
60000/60000 [==============================] - 4s 68us/sample - loss: 0.3019 - acc: 0.9226
Epoch 5/5
10000/10000 [==============================] - 0s 37us/sample - loss: 0.2736 - acc: 0.9266


In [ ]:
predictions = model.predict(x_test)
print("test accuracy : ", test_acc)

test accuracy :  0.9355


### With increase in the no of neurons there is slight increase in the accuracy

### 5th test -> increase in the no of epochs with 180 neuron

In [ ]:
model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(180, activation=tf.nn.relu),
    keras.layers.Dense(10, activation=tf.nn.softmax)
    ])

model.compile(optimizer=tf.train.AdamOptimizer(),
             loss='sparse_categorical_crossentropy',
             metrics=['accuracy']
             )

model.fit(x_train, y_train, epochs=9)
test_loss, test_acc = model.evaluate(x_test, y_test)
predictions = model.predict(x_test)

Train on 60000 samples
Epoch 1/9
60000/60000 [==============================] - 5s 91us/sample - loss: 2.8810 - acc: 0.8794
Epoch 2/9
60000/60000 [==============================] - 6s 92us/sample - loss: 0.3494 - acc: 0.9216
Epoch 3/9
60000/60000 [==============================] - 5s 86us/sample - loss: 0.2722 - acc: 0.9349
Epoch 4/9
60000/60000 [==============================] - 5s 83us/sample - loss: 0.2449 - acc: 0.9419
Epoch 5/9
60000/60000 [==============================] - 5s 84us/sample - loss: 0.2169 - acc: 0.9464
Epoch 6/9
60000/60000 [==============================] - 5s 85us/sample - loss: 0.2117 - acc: 0.9493
Epoch 7/9
60000/60000 [==============================] - 5s 86us/sample - loss: 0.1964 - acc: 0.9530
Epoch 8/9
60000/60000 [==============================] - 5s 85us/sample - loss: 0.1840 - acc: 0.9567
Epoch 9/9
10000/10000 [==============================] - 0s 48us/sample - loss: 0.3171 - acc: 0.9449


In [ ]:
print("test accuracy : ", test_acc)

test accuracy :  0.9468


### 6th test -> using the Sigmoid actiovation function in the layer

In [ ]:
model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(128, activation=tf.nn.sigmoid),
    keras.layers.Dense(10, activation=tf.nn.softmax)
    ])

model.compile(optimizer=tf.train.AdamOptimizer(),
             loss='sparse_categorical_crossentropy',
             metrics=['accuracy']
             )

model.fit(x_train, y_train, epochs=4)
test_loss, test_acc = model.evaluate(x_test, y_test)
predictions = model.predict(x_test)


Train on 60000 samples
Epoch 1/4
60000/60000 [==============================] - 5s 81us/sample - loss: 0.5284 - acc: 0.8560
Epoch 2/4
60000/60000 [==============================] - 5s 81us/sample - loss: 0.3500 - acc: 0.8981
Epoch 3/4
60000/60000 [==============================] - 5s 78us/sample - loss: 0.3200 - acc: 0.9054
Epoch 4/4
10000/10000 [==============================] - 0s 43us/sample - loss: 0.2851 - acc: 0.9131


In [ ]:
print("test accuracy : ", test_acc)

test accuracy :  0.9211


### There is a decrease in the accuracy with sigmoid activation function when compared with ReLU activation functin

### 7th test -> using the tanh activation functuion

In [ ]:
model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(128, activation=tf.nn.tanh),
    keras.layers.Dense(10, activation=tf.nn.softmax)
    ])

model.compile(optimizer=tf.trainAdamOptimizer(),
             loss='sparse_categorical_crossentropy',
             metrics=['accuracy']
             )

model.fit(x_train, y_train, epochs=3)
test_loss, test_acc = model.evaluate(x_test, y_test)
predictions = model.predict(x_test)


In [ ]:
print("test accuracy : ", test_acc)

test accuracy :  0.9131


### With Tanh activation function there is an observable decrease in the accuracy 

### 8th test -> using the softmax function in the hidden layer

In [ ]:
model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(128, activation=tf.nn.softmax),
    keras.layers.Dense(10, activation=tf.nn.softmax)
    ])

model.compile(optimizer=tf.train.AdamOptimizer(),
             loss='sparse_categorical_crossentropy',
             metrics=['accuracy']
             )

model.fit(x_train, y_train, epochs=2)
test_loss, test_acc = model.evaluate(x_test, y_test)
predictions = model.predict(x_test)

Train on 60000 samples
Epoch 1/2
60000/60000 [==============================] - 6s 100us/sample - loss: 1.6700 - acc: 0.4854
Epoch 2/2
10000/10000 [==============================] - 0s 44us/sample - loss: 1.1384 - acc: 0.5293


In [ ]:
print("test accuracy : ", test_acc)

test accuracy :  0.5424


### Softmax function in hidden caused a significant decrease in the accuracy

### 9th test -> using Stochastic gradient descent(SGD) optimizer

In [ ]:
model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(128, activation=tf.nn.relu),
    keras.layers.Dense(10, activation=tf.nn.softmax)
    ])

model.compile(optimizer=tf.train.AdamOptimizer(),
             loss='sparse_categorical_crossentropy',
             metrics=['accuracy']
             )

model.fit(x_train, y_train, epochs=10)
test_loss, test_acc = model.evaluate(x_test, y_test)
predictions = model.predict(x_test)

Train on 60000 samples
Epoch 1/10
60000/60000 [==============================] - 5s 90us/sample - loss: 2.5754 - acc: 0.8662
Epoch 2/10
60000/60000 [==============================] - 5s 79us/sample - loss: 0.3525 - acc: 0.9145
Epoch 3/10
60000/60000 [==============================] - 5s 77us/sample - loss: 0.2659 - acc: 0.9316
Epoch 4/10
60000/60000 [==============================] - 5s 77us/sample - loss: 0.2416 - acc: 0.9387
Epoch 5/10
60000/60000 [==============================] - 5s 79us/sample - loss: 0.2155 - acc: 0.9453
Epoch 6/10
60000/60000 [==============================] - 5s 81us/sample - loss: 0.2120 - acc: 0.9481
Epoch 7/10
60000/60000 [==============================] - 5s 78us/sample - loss: 0.1941 - acc: 0.9514
Epoch 8/10
60000/60000 [==============================] - 5s 79us/sample - loss: 0.1860 - acc: 0.9540
Epoch 9/10
60000/60000 [==============================] - 5s 76us/sample - loss: 0.1902 - acc: 0.9534
Epoch 10/10
10000/10000 [==============================] - 

In [ ]:
print("test accuracy : ", test_acc)

test accuracy :  0.9449


### SGD optimizer has the lowest accuracy when compared with Adam's optimizer

### 10th test -> using Adamax optimizer

In [ ]:
print("test accuracy : ", test_acc)

test accuracy :  0.9536


In [ ]:
model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(128, activation=tf.nn.relu),
    keras.layers.Dense(10, activation=tf.nn.softmax)
    ])

model.compile(optimizer=tf.train.AdamOptimizer(),
             loss='sparse_categorical_crossentropy',
             metrics=['accuracy']
             )

model.fit(x_train, y_train, epochs=18)
test_loss, test_acc = model.evaluate(x_test, y_test)
predictions = model.predict(x_test)

Train on 60000 samples
Epoch 1/18
60000/60000 [==============================] - 4s 65us/sample - loss: 2.8045 - acc: 0.8484
Epoch 2/18
60000/60000 [==============================] - 4s 64us/sample - loss: 0.3886 - acc: 0.9062
Epoch 3/18
60000/60000 [==============================] - 4s 64us/sample - loss: 0.2782 - acc: 0.9277
Epoch 4/18
60000/60000 [==============================] - 4s 64us/sample - loss: 0.2490 - acc: 0.9364
Epoch 5/18
60000/60000 [==============================] - 4s 63us/sample - loss: 0.2242 - acc: 0.9433
Epoch 6/18
60000/60000 [==============================] - 4s 63us/sample - loss: 0.2128 - acc: 0.9467
Epoch 7/18
60000/60000 [==============================] - 4s 65us/sample - loss: 0.2053 - acc: 0.9497
Epoch 8/18
60000/60000 [==============================] - 4s 64us/sample - loss: 0.1953 - acc: 0.9521
Epoch 9/18
60000/60000 [==============================] - 4s 63us/sample - loss: 0.1862 - acc: 0.9539
Epoch 10/18
60000/60000 [==============================] - 

### With Adamax optimizer these is a slight increase in the accuracy

In [ ]:
model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(128, activation=tf.nn.relu),
    keras.layers.Dense(10, activation=tf.nn.sigmoid)
    ])

model.compile(optimizer=tf.train.AdamOptimizer(),
             loss='sparse_categorical_crossentropy',
             metrics=['accuracy']
             )

model.fit(x_train, y_train, epochs=18)
test_loss, test_acc = model.evaluate(x_test, y_test)
predictions = model.predict(x_test)

Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where
Train on 60000 samples
Epoch 1/18
60000/60000 [==============================] - 4s 67us/sample - loss: 2.2072 - acc: 0.2126
Epoch 2/18
60000/60000 [==============================] - 4s 65us/sample - loss: 1.9936 - acc: 0.2871
Epoch 3/18
60000/60000 [==============================] - 4s 64us/sample - loss: 1.9411 - acc: 0.2930
Epoch 4/18
60000/60000 [==============================] - 4s 66us/sample - loss: 1.9172 - acc: 0.3357
Epoch 5/18
60000/60000 [==============================] - 4s 66us/sample - loss: 1.9589 - acc: 0.3661
Epoch 6/18
60000/60000 [==============================] - 4s 65us/sample - loss: 1.9738 - acc: 0.3738
Epoch 7/18
60000/60000 [==============================] - 4s 65us/sample - loss: 1.9225 - acc: 0.3489
Epoch 8/18
60000/60000 [==============================] - 4s 64us/sample - loss: 1.8759 - acc: 0.4112
Epoch 9/18
60000/60000 [==============================] - 4s 64us/s

In [ ]:
print("test accuracy : ", test_acc)

test accuracy :  0.5286


In [ ]:
model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(128, activation=tf.nn.relu),
    keras.layers.Dense(128, activation=tf.nn.relu),
    keras.layers.Dense(10, activation=tf.nn.softmax)
    ])

model.compile(optimizer=tf.train.AdamOptimizer(),
             loss='sparse_categorical_crossentropy',
             metrics=['accuracy']
             )

model.fit(x_train, y_train, epochs=18)
test_loss, test_acc = model.evaluate(x_test, y_test)
predictions = model.predict(x_test)

Train on 60000 samples
Epoch 1/18
60000/60000 [==============================] - 4s 74us/sample - loss: 1.7267 - acc: 0.8758
Epoch 2/18
60000/60000 [==============================] - 4s 74us/sample - loss: 0.3636 - acc: 0.9285
Epoch 3/18
60000/60000 [==============================] - 4s 74us/sample - loss: 0.2347 - acc: 0.9429
Epoch 4/18
60000/60000 [==============================] - 4s 74us/sample - loss: 0.1697 - acc: 0.9543
Epoch 5/18
60000/60000 [==============================] - 4s 74us/sample - loss: 0.1419 - acc: 0.9603
Epoch 6/18
60000/60000 [==============================] - 4s 73us/sample - loss: 0.1306 - acc: 0.9638
Epoch 7/18
60000/60000 [==============================] - 4s 73us/sample - loss: 0.1150 - acc: 0.9678
Epoch 8/18
60000/60000 [==============================] - 5s 77us/sample - loss: 0.1135 - acc: 0.9686
Epoch 9/18
60000/60000 [==============================] - 5s 76us/sample - loss: 0.1012 - acc: 0.9722
Epoch 10/18
60000/60000 [==============================] - 

In [ ]:
print("test accuracy : ", test_acc)

test accuracy :  0.9626
